File: web/console/scripts/task-transport.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { consoleTaskConversation } from "../src/lib/task-transport.ts";

test("task controls use explicit transport and preserve opaque conversation IDs", () => {
    assert.equal(consoleTaskConversation({ transport: "feishu", channel: "console:opaque-native-id" }), null);
    assert.equal(consoleTaskConversation({ channel: "console:unknown" }), null);
    assert.equal(consoleTaskConversation({ transport: "", channel: "console:unknown" }), null);
    assert.equal(consoleTaskConversation({ transport: "console", channel: "" }), null);
    assert.equal(consoleTaskConversation({ transport: "console" }), null);
    assert.equal(consoleTaskConversation({ transport: "console", channel: "opaque-without-prefix" }), "opaque-without-prefix");
    assert.equal(consoleTaskConversation({ transport: "console", channel: "console:main" }), "console:main");
});

File: internal/readmodel/task_transport_test.go
package readmodel

import (
	"encoding/json"
	"testing"

	"github.com/gopact-ai/steve/internal/task"
)

func TestTaskProjectionPreservesExplicitTransport(t *testing.T) {
	for _, tc := range []struct{ transport, conversation string }{
		{"feishu", "console:opaque-native-id"},
		{"console", "opaque-without-prefix"},
		{"", "console:unknown"},
	} {
		t.Run(tc.transport, func(t *testing.T) {
			projected := tasks([]task.Task{{ID: "fixture", Transport: tc.transport, Channel: tc.conversation}}, nil)
			raw, err := json.Marshal(projected[0])
			if err != nil {
				t.Fatal(err)
			}
			var wire map[string]any
			if err := json.Unmarshal(raw, &wire); err != nil {
				t.Fatal(err)
			}
			if got, _ := wire["transport"].(string); got != tc.transport {
				t.Fatalf("transport must be projected explicitly, never inferred from channel: %s", raw)
			}
			if wire["channel"] != tc.conversation {
				t.Fatalf("opaque conversation changed: %s", raw)
			}
		})
	}
}
